timemachine/fe/free_energy.py

```diff
@@ -725,3 +722,184 @@ def compute_intermediate_result(lambdas: Sequence[float]) -> PairBarResult:
+
+    for iteration, n_frames_iter in enumerate(batches(md_params.n_frames, n_frames_per_iter), 1):
+
+        def sample_replica(xb: CoordsBox, state_idx: StateIdx) -> Tuple[StoredArrays, NDArray]:
+            initial_state = replace(initial_states[state_idx], x0=xb.coords, box0=xb.box)
+            md_params_replica = replace(
+                md_params, n_frames=n_frames_iter, n_eq_steps=0, seed=np.random.randint(np.iinfo(np.int32).max)
+            )
+            frames, boxes = sample(initial_state, md_params_replica, max_buffer_frames=100)
+            return frames, boxes
+
+        def replica_from_samples(samples: Tuple[StoredArrays, NDArray]) -> CoordsBox:
+            frames, boxes = samples
+            return CoordsBox(frames[-1], boxes[-1])
+
+        hrex, samples_by_state = hrex.sample_replicas(sample_replica, replica_from_samples)
+        log_q = get_log_q_fn(replicas)
+        hrex, fraction_accepted_by_pair = hrex.attempt_neighbor_swaps(neighbor_pairs, log_q, n_swap_attempts_per_iter)
+        fraction_accepted_by_pair = fraction_accepted_by_pair[1:]  # remove stats for (0, 0) pair
+
+        samples_by_state_by_iter.append(samples_by_state)
+        replica_idx_by_state_by_iter.append(hrex.replica_idx_by_state)
+        fraction_accepted_by_pair_by_iter.append(fraction_accepted_by_pair)
+
+        if verbose:
+
+            def get_swap_acceptance_rates(fraction_accepted_by_pair):
+                return [
+                    n_accepted / n_proposed if n_proposed else np.nan
+                    for n_accepted, n_proposed in fraction_accepted_by_pair
+                ]
+
+            instantaneous_swap_acceptance_rates = get_swap_acceptance_rates(fraction_accepted_by_pair)
+            average_swap_acceptance_rates = get_swap_acceptance_rates(np.sum(fraction_accepted_by_pair_by_iter, axis=0))
+
+            def format_rate(r):
+                return f"{r * 100.0:5.1f}%"
+
+            def format_rates(rs):
+                return " | ".join(format_rate(r) for r in rs)
+
+            print("Completed HREX iteration", iteration)
+            print("Acceptance rates (inst.)   :", format_rates(instantaneous_swap_acceptance_rates))
+            print("Acceptance rates (average) :", format_rates(average_swap_acceptance_rates))
+            print("Final replica permutation  :", hrex.replica_idx_by_state)
+            print()
+
+    # Concatenate frames and boxes from all iterations
+    frames_by_state: List[StoredArrays] = [StoredArrays() for _ in lambdas]
+    boxes_by_state_: List[List[NDArray]] = [[] for _ in lambdas]
+    for samples_by_state in samples_by_state_by_iter:
+        for samples, frames, boxes in zip(samples_by_state, frames_by_state, boxes_by_state_):
+            frames_i, boxes_i = samples
+            frames.extend(frames_i)
+            boxes.extend(boxes_i)
+
+    boxes_by_state: List[NDArray] = [np.array(boxes) for boxes in boxes_by_state_]
+
+    energy_decomposed_states = [
```

Comment:
It looks like here you construct all the potentials in memory which could cause an OOM. Seems better to do it iteratively like we do in bisection

Reply:
Was fixed in https://github.com/proteneer/timemachine/pull/1128/commits/001202da338e20c48fcad0c9499da3da82cca19a and https://github.com/proteneer/timemachine/pull/1128/commits/6ff38e4d7bf6d66595a8c5b451f1ac4abde8404b